File: android/src/main/java/com/aspect/webviewbridge/modules/StatusBarModule.kt
/**
 * StatusBar 模块 - 状态栏控制
 *
 * 提供状态栏样式、可见性等控制功能
 */

package com.aspect.webviewbridge.modules

import android.app.Activity
import android.content.Context
import android.graphics.Color
import android.os.Build
import android.view.View
import android.view.WindowInsetsController
import android.view.WindowManager
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat
import com.aspect.webviewbridge.protocol.*

/**
 * 状态栏样式
 */
enum class StatusBarStyleType {
    DEFAULT,
    LIGHT,
    DARK
}

/**
 * 状态栏动画类型
 */
enum class StatusBarAnimation {
    NONE,
    FADE,
    SLIDE
}

/**
 * StatusBar 模块
 */
class StatusBarModule(
    private val context: Context,
    private val bridgeContext: BridgeModuleContext
) : BridgeModule {
    
    override val moduleName = "StatusBar"
    
    override val methods = listOf(
        "SetStyle",
        "SetVisible",
        "GetInfo",
        "SetBackgroundColor",
        "SetOverlaysWebView"
    )
    
    // 状态栏配置
    private var currentStyle = StatusBarStyleType.DEFAULT
    private var isVisible = true
    private var backgroundColor: Int = Color.TRANSPARENT
    private var overlaysWebView = true
    
    override fun handleRequest(
        method: String,
        request: BridgeRequest,
        callback: (Result<Any?>) -> Unit
    ) {
        when (method) {
            "SetStyle" -> setStyle(request, callback)
            "SetVisible" -> setVisible(request, callback)
            "GetInfo" -> getInfo(callback)
            "SetBackgroundColor" -> setBackgroundColor(request, callback)
            "SetOverlaysWebView" -> setOverlaysWebView(request, callback)
            else -> callback(Result.failure(BridgeError.methodNotFound("$moduleName.$method")))
        }
    }
    
    // MARK: - SetStyle
    
    private fun setStyle(
        request: BridgeRequest,
        callback: (Result<Any?>) -> Unit
    ) {
        val styleString = request.getString("style")
        if (styleString == null) {
            callback(Result.failure(BridgeError.invalidParams("style")))
            return
        }
        
        val style = try {
            StatusBarStyleType.valueOf(styleString.uppercase())
        } catch (e: Exception) {
            StatusBarStyleType.DEFAULT
        }
        
        val activity = context as? Activity
        if (activity == null) {
            callback(Result.failure(BridgeError.notSupported("需要 Activity 上下文")))
            return
        }
        
        try {
            currentStyle = style
            
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                val insetsController = activity.window.insetsController
                when (style) {
                    StatusBarStyleType.LIGHT -> {
                        // 浅色状态栏 = 深色图标
                        insetsController?.setSystemBarsAppearance(
                            WindowInsetsController.APPEARANCE_LIGHT_STATUS_BARS,
                            WindowInsetsController.APPEARANCE_LIGHT_STATUS_BARS
                        )
                    }
                    StatusBarStyleType.DARK -> {
                        // 深色状态栏 = 浅色图标
                        insetsController?.setSystemBarsAppearance(
                            0,
                            WindowInsetsController.APPEARANCE_LIGHT_STATUS_BARS
                        )
                    }
                    StatusBarStyleType.DEFAULT -> {
                        // 默认
                        insetsController?.setSystemBarsAppearance(
                            WindowInsetsController.APPEARANCE_LIGHT_STATUS_BARS,
                            WindowInsetsController.APPEARANCE_LIGHT_STATUS_BARS
                        )
                    }
                }
            } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                val decorView = activity.window.decorView
                @Suppress("DEPRECATION")
                var flags = decorView.systemUiVisibility
                
                flags = when (style) {
                    StatusBarStyleType.LIGHT -> {
                        flags or View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR
                    }
                    StatusBarStyleType.DARK -> {
                        flags and View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR.inv()
                    }
                    StatusBarStyleType.DEFAULT -> {
                        flags or View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR
                    }
                }
                
                @Suppress("DEPRECATION")
                decorView.systemUiVisibility = flags
            }
            
            callback(Result.success(null))
        } catch (e: Exception) {
            callback(Result.failure(BridgeError.internalError(e.message)))
        }
    }
    
    // MARK: - SetVisible
    
    private fun setVisible(
        request: BridgeRequest,
        callback: (Result<Any?>) -> Unit
    ) {
        val visible = request.getBool("visible")
        if (visible == null) {
            callback(Result.failure(BridgeError.invalidParams("visible")))
            return
        }
        
        val activity = context as? Activity
        if (activity == null) {
            callback(Result.failure(BridgeError.notSupported("需要 Activity 上下文")))
            return
        }
        
        try {
            isVisible = visible
            
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                val insetsController = activity.window.insetsController
                if (visible) {
                    insetsController?.show(android.view.WindowInsets.Type.statusBars())
                } else {
                    insetsController?.hide(android.view.WindowInsets.Type.statusBars())
                }
            } else {
                @Suppress("DEPRECATION")
                if (visible) {
                    activity.window.clearFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN)
                } else {
                    activity.window.addFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN)
                }
            }
            
            callback(Result.success(null))
        } catch (e: Exception) {
            callback(Result.failure(BridgeError.internalError(e.message)))
        }
    }
    
    // MARK: - GetInfo
    
    private fun getInfo(callback: (Result<Any?>) -> Unit) {
        val activity = context as? Activity
        
        var height = 0
        if (activity != null) {
            val resourceId = context.resources.getIdentifier(
                "status_bar_height",
                "dimen",
                "android"
            )
            if (resourceId > 0) {
                height = context.resources.getDimensionPixelSize(resourceId)
            }
        }
        
        val styleString = when (currentStyle) {
            StatusBarStyleType.LIGHT -> "light"
            StatusBarStyleType.DARK -> "dark"
            StatusBarStyleType.DEFAULT -> "default"
        }
        
        callback(Result.success(mapOf(
            "style" to styleString,
            "visible" to isVisible,
            "height" to height,
            "overlaysWebView" to overlaysWebView,
            "backgroundColor" to colorToHex(backgroundColor)
        )))
    }
    
    // MARK: - SetBackgroundColor
    
    private fun setBackgroundColor(
        request: BridgeRequest,
        callback: (Result<Any?>) -> Unit
    ) {
        val colorString = request.getString("color")
        if (colorString == null) {
            callback(Result.failure(BridgeError.invalidParams("color")))
            return
        }
        
        val activity = context as? Activity
        if (activity == null) {
            callback(Result.failure(BridgeError.notSupported("需要 Activity 上下文")))
            return
        }
        
        try {
            val color = parseColor(colorString)
            if (color == null) {
                callback(Result.failure(BridgeError.invalidParams("无效的颜色值")))
                return
            }
            
            backgroundColor = color
            
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                activity.window.statusBarColor = color
            }
            
            callback(Result.success(null))
        } catch (e: Exception) {
            callback(Result.failure(BridgeError.internalError(e.message)))
        }
    }
    
    // MARK: - SetOverlaysWebView
    
    private fun setOverlaysWebView(
        request: BridgeRequest,
        callback: (Result<Any?>) -> Unit
    ) {
        val overlay = request.getBool("overlay")
        if (overlay == null) {
            callback(Result.failure(BridgeError.invalidParams("overlay")))
            return
        }
        
        val activity = context as? Activity
        if (activity == null) {
            callback(Result.failure(BridgeError.notSupported("需要 Activity 上下文")))
            return
        }
        
        try {
            overlaysWebView = overlay
            
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                if (overlay) {
                    activity.window.addFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS)
                    WindowCompat.setDecorFitsSystemWindows(activity.window, false)
                } else {
                    activity.window.clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS)
                    WindowCompat.setDecorFitsSystemWindows(activity.window, true)
                }
            }
            
            callback(Result.success(null))
        } catch (e: Exception) {
            callback(Result.failure(BridgeError.internalError(e.message)))
        }
    }
    
    // MARK: - 辅助方法
    
    private fun parseColor(colorString: String): Int? {
        return try {
            Color.parseColor(colorString)
        } catch (e: Exception) {
            null
        }
    }
    
    private fun colorToHex(color: Int): String {
        val alpha = Color.alpha(color)
        val red = Color.red(color)
        val green = Color.green(color)
        val blue = Color.blue(color)
        
        return if (alpha == 255) {
            String.format("#%02X%02X%02X", red, green, blue)
        } else {
            String.format("#%02X%02X%02X%02X", alpha, red, green, blue)
        }
    }
}

File: webview-bridge-sdk/ios/Sources/Modules/StatusBarModule.swift
/**
 * StatusBar 模块 - 状态栏控制
 *
 * 提供状态栏样式、可见性等控制功能
 */

import Foundation
import UIKit

// MARK: - 状态栏样式

/// 状态栏样式
public enum StatusBarStyleType: String {
    case `default` = "default"
    case light = "light"
    case dark = "dark"
}

/// 状态栏动画类型
public enum StatusBarAnimation: String {
    case none
    case fade
    case slide
}

// MARK: - StatusBar 模块

public class StatusBarModule: BridgeModule {
    
    public let moduleName = "StatusBar"
    public let methods = ["SetStyle", "SetVisible", "GetInfo", "SetBackgroundColor", "SetOverlaysWebView"]
    
    private weak var bridge: WebViewBridge?
    
    // 状态栏配置
    private var currentStyle: UIStatusBarStyle = .default
    private var isHidden: Bool = false
    private var backgroundColor: UIColor?
    private var overlaysWebView: Bool = true
    
    public init(bridge: WebViewBridge) {
        self.bridge = bridge
    }
    
    public func handleRequest(
        method: String,
        params: [String: AnyCodable],
        callback: @escaping (Result<Any?, BridgeError>) -> Void
    ) {
        switch method {
        case "SetStyle":
            setStyle(params: params, callback: callback)
        case "SetVisible":
            setVisible(params: params, callback: callback)
        case "GetInfo":
            getInfo(callback: callback)
        case "SetBackgroundColor":
            setBackgroundColor(params: params, callback: callback)
        case "SetOverlaysWebView":
            setOverlaysWebView(params: params, callback: callback)
        default:
            callback(.failure(BridgeError.methodNotFound("\(moduleName).\(method)")))
        }
    }
    
    // MARK: - SetStyle
    
    private func setStyle(
        params: [String: AnyCodable],
        callback: @escaping (Result<Any?, BridgeError>) -> Void
    ) {
        guard let styleString = params.getString("style"),
              let style = StatusBarStyleType(rawValue: styleString) else {
            callback(.failure(BridgeError.invalidParams("style")))
            return
        }
        
        let animated = params.getBool("animated") ?? true
        
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            
            let newStyle: UIStatusBarStyle
            switch style {
            case .default:
                newStyle = .default
            case .light:
                newStyle = .lightContent
            case .dark:
                if #available(iOS 13.0, *) {
                    newStyle = .darkContent
                } else {
                    newStyle = .default
                }
            }
            
            self.currentStyle = newStyle
            
            // 通知视图控制器更新状态栏
            self.notifyStatusBarStyleChange(style: newStyle, animated: animated)
            
            callback(.success(nil))
        }
    }
    
    // MARK: - SetVisible
    
    private func setVisible(
        params: [String: AnyCodable],
        callback: @escaping (Result<Any?, BridgeError>) -> Void
    ) {
        guard let visible = params.getBool("visible") else {
            callback(.failure(BridgeError.invalidParams("visible")))
            return
        }
        
        let animationString = params.getString("animation") ?? "fade"
        let animation = StatusBarAnimation(rawValue: animationString) ?? .fade
        
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            
            self.isHidden = !visible
            
            let uiAnimation: UIStatusBarAnimation
            switch animation {
            case .none:
                uiAnimation = .none
            case .fade:
                uiAnimation = .fade
            case .slide:
                uiAnimation = .slide
            }
            
            self.notifyStatusBarVisibilityChange(hidden: !visible, animation: uiAnimation)
            
            callback(.success(nil))
        }
    }
    
    // MARK: - GetInfo
    
    private func getInfo(callback: @escaping (Result<Any?, BridgeError>) -> Void) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            
            let style: String
            switch self.currentStyle {
            case .lightContent:
                style = "light"
            case .default:
                style = "default"
            default:
                if #available(iOS 13.0, *), self.currentStyle == .darkContent {
                    style = "dark"
                } else {
                    style = "default"
                }
            }
            
            var height: CGFloat = 0
            if #available(iOS 13.0, *) {
                if let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene {
                    height = windowScene.statusBarManager?.statusBarFrame.height ?? 0
                }
            } else {
                height = UIApplication.shared.statusBarFrame.height
            }
            
            callback(.success([
                "style": style,
                "visible": !self.isHidden,
                "height": height,
                "overlaysWebView": self.overlaysWebView,
                "backgroundColor": self.backgroundColor?.hexString ?? NSNull()
            ]))
        }
    }
    
    // MARK: - SetBackgroundColor
    
    private func setBackgroundColor(
        params: [String: AnyCodable],
        callback: @escaping (Result<Any?, BridgeError>) -> Void
    ) {
        guard let colorString = params.getString("color") else {
            callback(.failure(BridgeError.invalidParams("color")))
            return
        }
        
        guard let color = UIColor(hex: colorString) else {
            callback(.failure(BridgeError.invalidParams("无效的颜色值")))
            return
        }
        
        DispatchQueue.main.async { [weak self] in
            self?.backgroundColor = color
            self?.notifyBackgroundColorChange(color: color)
            callback(.success(nil))
        }
    }
    
    // MARK: - SetOverlaysWebView
    
    private func setOverlaysWebView(
        params: [String: AnyCodable],
        callback: @escaping (Result<Any?, BridgeError>) -> Void
    ) {
        guard let overlay = params.getBool("overlay") else {
            callback(.failure(BridgeError.invalidParams("overlay")))
            return
        }
        
        DispatchQueue.main.async { [weak self] in
            self?.overlaysWebView = overlay
            self?.notifyOverlayChange(overlay: overlay)
            callback(.success(nil))
        }
    }
    
    // MARK: - 通知方法
    
    /// 通知状态栏样式变化
    private func notifyStatusBarStyleChange(style: UIStatusBarStyle, animated: Bool) {
        // 发送通知，让宿主应用处理
        NotificationCenter.default.post(
            name: .statusBarStyleDidChange,
            object: nil,
            userInfo: [
                "style": style,
                "animated": animated
            ]
        )
    }
    
    /// 通知状态栏可见性变化
    private func notifyStatusBarVisibilityChange(hidden: Bool, animation: UIStatusBarAnimation) {
        NotificationCenter.default.post(
            name: .statusBarVisibilityDidChange,
            object: nil,
            userInfo: [
                "hidden": hidden,
                "animation": animation
            ]
        )
    }
    
    /// 通知背景色变化
    private func notifyBackgroundColorChange(color: UIColor) {
        NotificationCenter.default.post(
            name: .statusBarBackgroundColorDidChange,
            object: nil,
            userInfo: ["color": color]
        )
    }
    
    /// 通知覆盖模式变化
    private func notifyOverlayChange(overlay: Bool) {
        NotificationCenter.default.post(
            name: .statusBarOverlayDidChange,
            object: nil,
            userInfo: ["overlay": overlay]
        )
    }
}

// MARK: - 通知名称扩展

public extension Notification.Name {
    static let statusBarStyleDidChange = Notification.Name("WebViewBridge.StatusBarStyleDidChange")
    static let statusBarVisibilityDidChange = Notification.Name("WebViewBridge.StatusBarVisibilityDidChange")
    static let statusBarBackgroundColorDidChange = Notification.Name("WebViewBridge.StatusBarBackgroundColorDidChange")
    static let statusBarOverlayDidChange = Notification.Name("WebViewBridge.StatusBarOverlayDidChange")
}

// MARK: - UIColor 扩展

private extension UIColor {
    /// 从十六进制字符串创建颜色
    convenience init?(hex: String) {
        var hexSanitized = hex.trimmingCharacters(in: .whitespacesAndNewlines)
        hexSanitized = hexSanitized.replacingOccurrences(of: "#", with: "")
        
        var rgb: UInt64 = 0
        guard Scanner(string: hexSanitized).scanHexInt64(&rgb) else {
            return nil
        }
        
        let length = hexSanitized.count
        
        var r, g, b, a: CGFloat
        switch length {
        case 6: // RGB
            r = CGFloat((rgb & 0xFF0000) >> 16) / 255.0
            g = CGFloat((rgb & 0x00FF00) >> 8) / 255.0
            b = CGFloat(rgb & 0x0000FF) / 255.0
            a = 1.0
        case 8: // ARGB
            a = CGFloat((rgb & 0xFF000000) >> 24) / 255.0
            r = CGFloat((rgb & 0x00FF0000) >> 16) / 255.0
            g = CGFloat((rgb & 0x0000FF00) >> 8) / 255.0
            b = CGFloat(rgb & 0x000000FF) / 255.0
        default:
            return nil
        }
        
        self.init(red: r, green: g, blue: b, alpha: a)
    }
    
    /// 转换为十六进制字符串
    var hexString: String {
        var r: CGFloat = 0
        var g: CGFloat = 0
        var b: CGFloat = 0
        var a: CGFloat = 0
        
        getRed(&r, green: &g, blue: &b, alpha: &a)
        
        if a < 1.0 {
            return String(
                format: "#%02X%02X%02X%02X",
                Int(a * 255),
                Int(r * 255),
                Int(g * 255),
                Int(b * 255)
            )
        } else {
            return String(
                format: "#%02X%02X%02X",
                Int(r * 255),
                Int(g * 255),
                Int(b * 255)
            )
        }
    }
}
